### ISSUE TITLE
retries when a node goes offline

### ISSUE THREAD

**Issue by amorton** [cassandra-rb/cassandra#94]
Using release 0.11.1, the retries argument does not work as I expect. It looks like a thrift thing but I'd thought I start here. 

Running against two local nodes 127.0.0.1 and 127.0.0.2, using irb...

``` ruby
>> cas = Cassandra.new('dev', "127.0.0.1:9160", :retries => 2, :connect_timeout => 5)
=> #<Cassandra:2160831680, @keyspace="dev", @schema={}, @servers=["127.0.0.1:9160"]>

# request goes to 127.0.0.2
>> cas.get(:data, "foo")
=> #<OrderedHash {"bar"=>"baz"}
{"bar"=>1308621620616000}>

#shutdown 127.0.0.2, no calls make it to the server logs
>> cas.get(:data, "foo")
ThriftClient::NoServersAvailable: No live servers in [#<AbstractThriftClient::Server:0x101971f38 @marked_down_at=Tue Jun 21 15:51:24 +1200 2011, @connection_string="127.0.0.1:9160">, #<AbstractThriftClient::Server:0x101971f10 @marked_down_at=Tue Jun 21 15:51:24 +1200 2011, @connection_string="127.0.0.2:9160">] since nil.
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:105:in `next_live_server'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:81:in `connect!'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:111:in `handled_proxy'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:58:in `multiget_slice'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/protocol.rb:83:in `_multiget'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/cassandra.rb:599:in `multi_get'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/cassandra.rb:572:in `get'
    from (irb):13
    from :0

# immediately exec the command again, read goes to 127.0.0.1
>> cas.get(:data, "foo")
=> #<OrderedHash {"bar"=>"baz"}
{"bar"=>1308621620616000}>
```

I also tried with retries set to 5. 

I also tried with using the exception_classes param

``` ruby
>> cas = Cassandra.new('dev', "127.0.0.1:9160", :retries => 5, :connect_timeout => 5, :exception_classes => [])
=> #<Cassandra:2160240660, @keyspace="dev", @schema={}, @servers=["127.0.0.1:9160"]>
>> 
?> 
?> cas.get(:data, "foo")
=> #<OrderedHash {"bar"=>"baz"}
{"bar"=>1308621620616000}>

# kill 127.0.0.2 where the last read happened 
>> cas.get(:data, "foo")
CassandraThrift::Cassandra::Client::TransportException: CassandraThrift::Cassandra::Client::TransportException
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/transport/socket.rb:120:in `read'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/transport/base_transport.rb:52:in `read_all'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/transport/framed_transport.rb:78:in `read_frame'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/transport/framed_transport.rb:49:in `read'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/transport/base_transport.rb:52:in `read_all'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/protocol/binary_protocol.rb:186:in `read_i32'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/protocol/binary_protocol.rb:111:in `read_message_begin'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/client.rb:45:in `receive_message'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/../vendor/0.7/gen-rb/cassandra.rb:112:in `recv_multiget_slice'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/../vendor/0.7/gen-rb/cassandra.rb:104:in `multiget_slice'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:116:in `send'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:116:in `handled_proxy'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:58:in `multiget_slice'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/protocol.rb:83:in `_multiget'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/cassandra.rb:599:in `multi_get'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/cassandra.rb:572:in `get'
    from (irb):36
    from :0

>>cas.get(:data, "foo")
IOError: closed stream
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/transport/socket.rb:62:in `write'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/transport/framed_transport.rb:70:in `flush'
    from /Library/Ruby/Gems/1.8/gems/thrift-0.6.0/lib/thrift/client.rb:41:in `send_message'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/../vendor/0.7/gen-rb/cassandra.rb:108:in `send_multiget_slice'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/../vendor/0.7/gen-rb/cassandra.rb:103:in `multiget_slice'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:116:in `send'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:116:in `handled_proxy'
    from /Library/Ruby/Gems/1.8/gems/thrift_client-0.6.2/lib/thrift_client/abstract_thrift_client.rb:58:in `multiget_slice'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/protocol.rb:83:in `_multiget'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/cassandra.rb:599:in `multi_get'
    from /Library/Ruby/Gems/1.8/gems/cassandra-0.11.1/lib/cassandra/cassandra.rb:572:in `get'
    from (irb):37
    from :0
# did not recover 
```

Whats the recommended approach to have the client handle nodes going down ? 
Also it looks like there may be a problem with thrift marking all the nodes down in the first example. I'm dont know a lot of ruby wanted to see if I was using the cassandra client wrong first.

Thanks

**Comment by mheffner** [comment]
It sounds like this is similar to what we were seeing. I believe the problem is that thrift_client will reconnect on an error and retry the command. However, Cassandra requires that the keyspace be selected before retrying the command.

We've been experimenting with [this commit](https://github.com/fauna/thrift_client/commit/a68a8cb6dfe39a9905f4527fc3c37007a4703632) that adds a callback to thrift_client after a connection (or reconnect). Similarly, we're using [this branch](https://github.com/mheffner/cassandra/tree/use_post_conn_cb) of 'fauna/cassandra' that leverages the callback to reset the keyspace and relogin before the command is retried.

**Comment by amorton** [comment]
ah, sounds plausible. I'll get to look at it again in a day or two. I'll check if the client is indeed making the second call but the keyspace has not been set. Cheers.

**Comment by ryanking** [comment]
aaron- let me know if that solves the issue and I can merge both forks.

**Comment by amorton** [comment]
Yup the existing code is making the request without setting the keyspace, service is throwing an InvalidRequestException that is been hidden on the client. 

``` java
InvalidRequestException(why:You have not set a keyspace for this session)
```

Worked as expected using https://github.com/fauna/thrift_client and the https://github.com/mheffner/cassandra/tree/use_post_conn_cb 

Client failed over to the other node and retried.

**Comment by coffeeaddict** [comment]
Hi!

I have a patched version of thrift_client I use for development right now where I get output like:

```
Handled exception with retry: CassandraThrift::InvalidRequestException: No indexed columns present in index clause with operator EQ
Handled exception with retry: CassandraThrift::InvalidRequestException: You have not set a keyspace for this session
Re-raised exception: No live servers in [192.168.2.1:9160, 192.168.2.4:9160] since nil.: No live servers in [192.168.2.1:9160, 192.168.2.4:9160] since nil.
ThriftClient::NoServersAvailable: No live servers in [192.168.2.1:9160, 192.168.2.4:9160] since nil.
```

The no live servers available message is very confusing when you are dabbling with indexes...

Here's the patch:

```
diff --git a/lib/thrift_client/abstract_thrift_client.rb b/lib/thrift_client/abstract_thrift_client.rb
index 07bf100..d7edfcb 100644
--- a/lib/thrift_client/abstract_thrift_client.rb
+++ b/lib/thrift_client/abstract_thrift_client.rb
@@ -114,6 +114,7 @@ class AbstractThriftClient
       @request_count += 1
       @client.send(method_name, *args)
     rescue *@options[:exception_classes] => e
+      $stderr.puts "Handled exception with retry: #{e}: #{e.message}"
       disconnect_on_error!
       tries ||= (@options[:retry_overrides][method_name.to_sym] || @options[:retries]) + 1
       tries -= 1
@@ -123,6 +124,7 @@ class AbstractThriftClient
         raise_or_default(e, method_name)
       end
     rescue Exception => e
+      $stderr.puts "Re-raised exception: #{e}: #{e.message}"
       raise_or_default(e, method_name)
     end
   end
```

Would it be possible to keep a stack of errors in the thrift_client (I'd be happy to provide a pull request for that) and make that stack actually accessible in Cassandra#client? This way we can see what the underlying problem is... I've glanced over the Cassandra source but there seems to be quite a number of delegations when it comes to the client

**Comment by ryanking** [comment]
I just released cassandra 0.11.2 that require thrift_client 0.6.3, which should fix this issue.

**Comment by mheffner** [comment]
The support for utilizing the post connection callback doesn't appear to have been merged into 'fauna/cassandra'. I've submitted a pull request for this feature here: https://github.com/fauna/cassandra/pull/96
